File: problem_solving_with_DSA/array/python/arrayRotate.py
"""
Rotate Array by K position
"""
def rightRotate(arr, n):
    """ Shift Array to the right side by 1 position
    
    Args:
        arr (arr[int]) - Array to rotate
        n (int): Size of the array
    """
    temp = arr[n-1]
    # for i in range(n-1, 0, -1):
    for i in reversed(range(1, n)):
        arr[i] = arr[i-1]
    arr[0] = temp
        

def arrayRotate(arr,n,k):
    """ Rotate array by k Position
    
    Args:
        arr (arr[int]) - Array to rotate
        n (int): Size of the array
        k (int): Rotation position
    """
    if n <= 1:
        return
    while k > 0:
        rightRotate(arr, n)
        k -= 1
        
def printArray(arr, n):
    """Print array
    
    Args:
        arr (arr[int]): array
        n (int) -  Size of the array
    """
    print('[', end="")
    for i in range(n):
        if i < n-1:
            print("{}, ".format(arr[i]), end="")
        else:
            print("{}]".format(arr[i]))

if __name__ == '__main__':
    arr = [1, 2, 3, 4, 5]
    len = len(arr)
    
    print("<<<Before rotation>>>")
    printArray(arr, len)
    
    k = int(input("Enter Rotation position\n"))
    arrayRotate(arr, len, k);
    
    print("<<<After rotation>>>")
    printArray(arr, len)

File: problem_solving_with_DSA/array/C/arrayRotate.c
#include <stdio.h>

/**
 * swap - Swap two integers 
 * 
 * @a: First integer 
 * @b: Second integer 
 */
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

/**
 * rotateArray - Rotate array by K position
 * 
 * @arr: Array to rotate
 * @n: Size of the array
 * @k: Number of times the array needs to be rotated
 */
void rotateArray(int arr[], int n, int k)
{
    if (n <= 1) return;
    
    while (k > 0)
    {
        for (int i = 0; i < n-1; i++)
        {
            swap(&arr[i], &arr[n-1]);
        }
        k--;
    }
}

/**
 * printArray - Print array 
 * 
 * @arr: Array to print
 * @n: Size of the array
 */
void printArray(int arr[], int n)
{
    printf("{");
    for (int i = 0; i < n; i++)
    {
        if (i < n-1)
            printf("%d, ", arr[i]);
        else
            printf("%d}\n", arr[i]);
    }
}


int main(void)
{
    int arr[] = {1, 2, 3, 4, 5};
    int size = sizeof(arr) / sizeof(arr[0]);
    int k;
    printf("Enter rotation number\n");
    scanf("%d", &k);
    
    printf("Before Rotation...\n");
    printArray(arr, size);
    
    rotateArray(arr, size, k);
    
    printf("After Rotation...\n");
    printArray(arr, size);
}